### ISSUE TITLE
Wrong branch coverage for Python 3.8 in try-return-finally

### ISSUE THREAD

**Issue by lhupfeldt** [coveragepy/coveragepy#707]
Given the following program:
```python
_val = None

def return_from_try_finaly(early):
    global _val
    _val = 1
    try:
        orig_val = _val
        _val = 2
        if early:
            return 1
    finally:
        _val = orig_val

    return 3


assert return_from_try_finaly(True) == 1
assert _val == 1

assert return_from_try_finaly(False) == 3
assert _val == 1
```
stored in `try_return_finally.py`

Shows full coverage for python 3.6:
```
$ python=python3.6; $python -m coverage run try_return_finally.py && $python -m coverage report -m
Name                    Stmts   Miss Branch BrPart      Cover   Missing
-----------------------------------------------------------------------
try_return_finally.py      14      0      4      0   100.000%
```

Partial branch for python 3.8:
```
$ python=python3.8; $python -m coverage run try_return_finally.py && $python -m coverage report -m
Name                    Stmts   Miss Branch BrPart      Cover   Missing
-----------------------------------------------------------------------
try_return_finally.py      14      0      4      1    94.444%   12->exit
```

**Comment by nedbat** [comment]
Hmm, looks like something has changed since 3.7...  Running the whole test suite on latest 3.8 shows these failures:
```
FAIL tests/test_arcs.py::ExceptionArcTest::test_break_through_finally
FAIL tests/test_arcs.py::ExceptionArcTest::test_else_jump_finally
FAIL tests/test_arcs.py::LoopArcTest::test_zero_coverage_while_loop
FAIL tests/test_arcs.py::ExceptionArcTest::test_continue_through_finally
FAIL tests/test_arcs.py::LoopArcTest::test_bug_496_continue_in_constant_while
FAIL tests/test_arcs.py::ExceptionArcTest::test_except_jump_finally
FAIL tests/test_arcs.py::ExceptionArcTest::test_return_finally
FAIL tests/test_arcs.py::LoopArcTest::test_while_true
FAIL tests/test_arcs.py::ExceptionArcTest::test_bug_212
FAIL tests/test_concurrency.py::ConcurrencyTest::test_threads
FAIL tests/test_html.py::HtmlGoldTests::test_partial
```

**Comment by nedbat** [comment]
I've asked the core devs for an opinion: https://bugs.python.org/issue34705

**Comment by lhupfeldt** [comment]
Hi again Ned,

In the latest version of Python 3.8 (11c4eaa99362f91c7faea88e31df3e46af020023) coverage is very broken, you can see here: https://travis-ci.org/lhupfeldt/multiconf/jobs/436314142.

I have 100% coverage on Python 3.6 and 3.7. All these coverage errors are new, they were not there when I created this issue with an earlier version of Python 3.8.

Sorry I don't have time to dig into it and provide examples right now.

You might want to alert core devs again.

**Comment by lhupfeldt** [comment]
multiconf/envs.py                 210     37     82      2    86.644%   313-1149, 91->125, 92->91
This file only has 307 lines, something has gone completely wrong.

**Comment by nedbat** [comment]
This is actually a different cause than the original problem.  The phantom lines are due to a change in the lnotab table that was made in Python 3.6, but hasn't been used until now.  The phantom line problem is a simple fix that I have to decide how to ship.

**Comment by nedbat** [comment]
The phantom line issue is #714

**Comment by nedbat** [comment]
@lhupfeldt I've made some fixes. Can you try installing from GitHub to see if it works for you?
```
pip install git+https://github.com/nedbat/coveragepy.git@nedbat/v4.5.x#egg=coverage==0.0
```

**Comment by lhupfeldt** [comment]
Both the try-finally and the phantom lines issues are solved. Thank you.

Using pip install with the link did not work though, I ran python3.8 setup.py install from commit 51aacdbe instead.

**Comment by lhupfeldt** [comment]
You had an extra 'nedbat' in pip install command, it should be:
pip install git+https://github.com/nedbat/coveragepy.git@v4.5.x#egg=coverage==0.0

**Comment by nedbat** [comment]
Oops, thanks!

**Comment by lhupfeldt** [comment]
No problem.

**Comment by nedbat** [comment]
Fixed in 04ff1883 (for 5.0) and f12f8559 (for 4.5.x)

**Comment by nedbat** [comment]
This fix is in 4.5.2, just released: https://pypi.org/project/coverage/4.5.2/
